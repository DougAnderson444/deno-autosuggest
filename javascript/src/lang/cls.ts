
export const ANY         = 0
export const CONTROL     = 1
export const WHITESPACE  = 2
export const PUNCTUATION = 3
export const NOTALPHA    = 4
export const NOTALPHANUM = 5
export const CONSONANT   = 6
export const VOWEL       = 7


export type CharClass =
    | typeof ANY
    | typeof CONTROL
    | typeof WHITESPACE
    | typeof PUNCTUATION
    | typeof NOTALPHA
    | typeof NOTALPHANUM
    | typeof CONSONANT
    | typeof VOWEL

