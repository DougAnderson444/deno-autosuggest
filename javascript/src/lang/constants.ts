import {CharClass, CONSONANT, VOWEL} from './cls'

export const CHAR_CLASSES_LATIN: [CharClass, string][] = [
    [CONSONANT, 'b'],
    [CONSONANT, 'c'],
    [CONSONANT, 'd'],
    [CONSONANT, 'f'],
    [CONSONANT, 'g'],
    [CONSONANT, 'h'],
    [CONSONANT, 'j'],
    [CONSONANT, 'k'],
    [CONSONANT, 'l'],
    [CONSONANT, 'm'],
    [CONSONANT, 'n'],
    [CONSONANT, 'p'],
    [CONSONANT, 'q'],
    [CONSONANT, 'r'],
    [CONSONANT, 's'],
    [CONSONANT, 't'],
    [CONSONANT, 'v'],
    [CONSONANT, 'w'],
    [CONSONANT, 'x'],
    [CONSONANT, 'z'],
    [VOWEL, 'a'],
    [VOWEL, 'e'],
    [VOWEL, 'i'],
    [VOWEL, 'o'],
    [VOWEL, 'u'],
    [VOWEL, 'y'],
]
