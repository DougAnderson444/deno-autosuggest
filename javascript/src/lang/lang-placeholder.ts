import {LangAbstract} from './lang'
import {Stemmer} from './stemming/ext/PlaceholderStemmer'

export class Lang extends LangAbstract {
    constructor() {
        super(
            new Stemmer(),
            [],
            [],
            [],
            [],
        )
    }
}
