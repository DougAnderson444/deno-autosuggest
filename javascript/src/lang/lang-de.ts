import {LangAbstract} from './lang'
import {Stemmer} from './stemming/ext/GermanStemmer'
import {CHAR_CLASSES_LATIN} from './constants'
import {CharClass, CONSONANT} from './cls'
import {
    PartOfSpeech,
    ARTICLE,
    PREPOSITION,
    CONJUNCTION,
    PARTICLE,
} from './pos'


const FUNCTION_WORDS: [PartOfSpeech, string][] = [
    [ARTICLE, "das"],
    [ARTICLE, "dem"],
    [ARTICLE, "den"],
    [ARTICLE, "der"],
    [ARTICLE, "des"],
    [ARTICLE, "die"],
    [ARTICLE, "ein"],
    [ARTICLE, "eine"],
    [ARTICLE, "einem"],
    [ARTICLE, "einen"],
    [ARTICLE, "einer"],
    [ARTICLE, "eines"],

    [PREPOSITION, "an"],
    [PREPOSITION, "auf"],
    [PREPOSITION, "aus"],
    [PREPOSITION, "bei"],
    [PREPOSITION, "bis"],
    [PREPOSITION, "durch"],
    [PREPOSITION, "entlang"],
    [PREPOSITION, "für"],
    [PREPOSITION, "gegen"],
    [PREPOSITION, "hinter"],
    [PREPOSITION, "in"],
    [PREPOSITION, "mit"],
    [PREPOSITION, "nach"],
    [PREPOSITION, "neben"],
    [PREPOSITION, "ohne"],
    [PREPOSITION, "seit"],
    [PREPOSITION, "um"],
    [PREPOSITION, "von"],
    [PREPOSITION, "zu"],

    [CONJUNCTION, "aber"],
    [CONJUNCTION, "als"],
    [CONJUNCTION, "als"],
    [CONJUNCTION, "anstatt"],
    [CONJUNCTION, "auch"],
    [CONJUNCTION, "bevor"],
    [CONJUNCTION, "bis"],
    [CONJUNCTION, "but"],
    [CONJUNCTION, "damit"],
    [CONJUNCTION, "dass"],
    [CONJUNCTION, "denn"],
    [CONJUNCTION, "entweder"],
    [CONJUNCTION, "nachdem"],
    [CONJUNCTION, "noch"],
    [CONJUNCTION, "ob"],
    [CONJUNCTION, "obwohl"],
    [CONJUNCTION, "oder"],
    [CONJUNCTION, "oder"],
    [CONJUNCTION, "seitdem"],
    [CONJUNCTION, "sobald"],
    [CONJUNCTION, "sofern"],
    [CONJUNCTION, "sondern"],
    [CONJUNCTION, "soweit"],
    [CONJUNCTION, "sowie"],
    [CONJUNCTION, "sowohl"],
    [CONJUNCTION, "sowohl"],
    [CONJUNCTION, "the"],
    [CONJUNCTION, "und"],
    [CONJUNCTION, "während"],
    [CONJUNCTION, "weder"],
    [CONJUNCTION, "weil"],
    [CONJUNCTION, "wenn"],
    [CONJUNCTION, "wie"],
    [CONJUNCTION, "wie"],
    [CONJUNCTION, "wo"],
    [CONJUNCTION, "zu"],

    [PARTICLE, "schon"],
    [PARTICLE, "ja"],
    [PARTICLE, "halt"],
    [PARTICLE, "wohl"],
    [PARTICLE, "doch"],
    [PARTICLE, "mal"],
    [PARTICLE, "aber"],
    [PARTICLE, "auch"],
    [PARTICLE, "bloß"],
    [PARTICLE, "denn"],
    [PARTICLE, "eben"],
    [PARTICLE, "etwas"],
    [PARTICLE, "nur"],
    [PARTICLE, "ruhig"],
    [PARTICLE, "shon"],
    [PARTICLE, "zwar"],
    [PARTICLE, "soweiso"],
]


const CHAR_CLASSES: [CharClass, string][] = [
    [CONSONANT, 'ß'],
]


const UTF_COMPOSE_MAP: [string, string][] = [
    ["Ä", "Ä"],
    ["Ö", "Ö"],
    ["Ü", "Ü"],
    ["ä", "ä"],
    ["ö", "ö"],
    ["ü", "ü"],
]


const UTF_REDUCE_MAP: [string, string][] = [
    ["ẞ", "SS"], // eszett
    ["ß", "ss"],
    ["Ä", "A"], // umlauts
    ["Ö", "O"],
    ["Ü", "U"],
    ["ä", "a"],
    ["ö", "o"],
    ["ü", "u"],
]


export class Lang extends LangAbstract {
    constructor() {
        super(
            new Stemmer(),
            [...CHAR_CLASSES_LATIN, ...CHAR_CLASSES],
            FUNCTION_WORDS,
            UTF_COMPOSE_MAP,
            UTF_REDUCE_MAP,
        )
    }
}
