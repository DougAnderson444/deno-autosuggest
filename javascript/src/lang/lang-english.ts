import {Lang} from './lang'
import {EnglishStemmer} from './stemming/ext/EnglishStemmer'
import {CHAR_CLASSES_LATIN} from './constants'
import {CharClass} from './cls'
import {
    PartOfSpeech,
    ARTICLE,
    PREPOSITION,
    CONJUNCTION,
    PARTICLE,
} from './pos'


const FUNCTION_WORDS: [PartOfSpeech, string][] = [
    [ARTICLE, "a"],
    [ARTICLE, "an"],
    [ARTICLE, "the"],

    [PREPOSITION, "at"],
    [PREPOSITION, "by"],
    [PREPOSITION, "for"],
    [PREPOSITION, "from"],
    [PREPOSITION, "in"],
    [PREPOSITION, "of"],
    [PREPOSITION, "on"],
    [PREPOSITION, "to"],
    [PREPOSITION, "since"],
    [PREPOSITION, "before"],
    [PREPOSITION, "till"],
    [PREPOSITION, "untill"],
    [PREPOSITION, "beside"],
    [PREPOSITION, "under"],
    [PREPOSITION, "below"],
    [PREPOSITION, "over"],
    [PREPOSITION, "above"],
    [PREPOSITION, "across"],
    [PREPOSITION, "through"],
    [PREPOSITION, "into"],
    [PREPOSITION, "towards"],
    [PREPOSITION, "onto"],
    [PREPOSITION, "off"],
    [PREPOSITION, "out"],
    [PREPOSITION, "about"],

    [CONJUNCTION, "after"],
    [CONJUNCTION, "although"],
    [CONJUNCTION, "as"],
    [CONJUNCTION, "because"],
    [CONJUNCTION, "before"],
    [CONJUNCTION, "but"],
    [CONJUNCTION, "either"],
    [CONJUNCTION, "for"],
    [CONJUNCTION, "how"],
    [CONJUNCTION, "if"],
    [CONJUNCTION, "lest"],
    [CONJUNCTION, "nor"],
    [CONJUNCTION, "once"],
    [CONJUNCTION, "once"],
    [CONJUNCTION, "or"],
    [CONJUNCTION, "since"],
    [CONJUNCTION, "since"],
    [CONJUNCTION, "since"],
    [CONJUNCTION, "so"],
    [CONJUNCTION, "than"],
    [CONJUNCTION, "that"],
    [CONJUNCTION, "that"],
    [CONJUNCTION, "though"],
    [CONJUNCTION, "till"],
    [CONJUNCTION, "unless"],
    [CONJUNCTION, "until"],
    [CONJUNCTION, "until"],
    [CONJUNCTION, "until"],
    [CONJUNCTION, "what"],
    [CONJUNCTION, "whatever"],
    [CONJUNCTION, "when"],
    [CONJUNCTION, "when"],
    [CONJUNCTION, "whenever"],
    [CONJUNCTION, "where"],
    [CONJUNCTION, "whereas"],
    [CONJUNCTION, "whereas"],
    [CONJUNCTION, "wherever"],
    [CONJUNCTION, "whether"],
    [CONJUNCTION, "which"],
    [CONJUNCTION, "whichever"],
    [CONJUNCTION, "while"],
    [CONJUNCTION, "while"],
    [CONJUNCTION, "while"],
    [CONJUNCTION, "whilst"],
    [CONJUNCTION, "who"],
    [CONJUNCTION, "whoever"],
    [CONJUNCTION, "whom"],
    [CONJUNCTION, "whomever"],
    [CONJUNCTION, "whose"],
    [CONJUNCTION, "why"],
    [CONJUNCTION, "yet"],
    [CONJUNCTION, "and"],
    // [CONJUNCTION, "as if"],
    // [CONJUNCTION, "as long as"],
    // [CONJUNCTION, "as much as"],
    // [CONJUNCTION, "as soon as"],
    // [CONJUNCTION, "as though"],
    // [CONJUNCTION, "assuming that"],
    // [CONJUNCTION, "by the time"],
    // [CONJUNCTION, "even if"],
    // [CONJUNCTION, "even though"],
    // [CONJUNCTION, "in case that"],
    // [CONJUNCTION, "in case"],
    // [CONJUNCTION, "in order that"],
    // [CONJUNCTION, "in order"],
    // [CONJUNCTION, "now that"],
    // [CONJUNCTION, "only if"],
    // [CONJUNCTION, "provided that"],
    // [CONJUNCTION, "rather than"],
    // [CONJUNCTION, "so that"],

    [PARTICLE, "by"],
    [PARTICLE, "in"],
    [PARTICLE, "not"],
    [PARTICLE, "on"],
    [PARTICLE, "to"],
    [PARTICLE, "oh"],
]


const CHAR_CLASSES: [CharClass, string][] = [
]


const UTF_COMPOSE_MAP: [string, string][] = [
]


const UTF_REDUCE_MAP: [string, string][] = [
]


export const LANG_ENGLISH = new Lang(
    new EnglishStemmer(),
    [...CHAR_CLASSES_LATIN, ...CHAR_CLASSES],
    FUNCTION_WORDS,
    UTF_COMPOSE_MAP,
    UTF_REDUCE_MAP,
)
