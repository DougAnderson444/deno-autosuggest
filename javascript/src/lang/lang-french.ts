import {Lang} from './lang'
import {FrenchStemmer} from './stemming/ext/FrenchStemmer'
import {CHAR_CLASSES_LATIN} from './constants'
import {CharClass, CONSONANT, VOWEL} from './cls'
import {
    PartOfSpeech,
    ARTICLE,
    PREPOSITION,
    CONJUNCTION,
    PARTICLE,
} from './pos'


const FUNCTION_WORDS: [PartOfSpeech, string][] = [
    [ARTICLE, "le"],
    [ARTICLE, "la"],
    [ARTICLE, "les"],
    [ARTICLE, "l"],
    [ARTICLE, "un"],
    [ARTICLE, "une"],
    [ARTICLE, "des"],
    [ARTICLE, "du"],
    [ARTICLE, "de"],

    [PREPOSITION, "à"],
    [PREPOSITION, "après"],
    [PREPOSITION, "au-delà"],
    [PREPOSITION, "au-dessus"],
    [PREPOSITION, "avant"],
    [PREPOSITION, "avec"],
    [PREPOSITION, "ces"],
    [PREPOSITION, "cette"],
    [PREPOSITION, "ceux"],
    [PREPOSITION, "comme"],
    [PREPOSITION, "contre"],
    [PREPOSITION, "dans"],
    [PREPOSITION, "de"],
    [PREPOSITION, "depuis"],
    [PREPOSITION, "derrière"],
    [PREPOSITION, "entre"],
    [PREPOSITION, "jusqu'à"],
    [PREPOSITION, "mais"],
    [PREPOSITION, "malgré"],
    [PREPOSITION, "opposé"],
    [PREPOSITION, "par"],
    [PREPOSITION, "plus"],
    [PREPOSITION, "pour"],
    [PREPOSITION, "prochain"],
    [PREPOSITION, "que"],
    [PREPOSITION, "qui"],
    [PREPOSITION, "sans"],
    [PREPOSITION, "sauf"],
    [PREPOSITION, "selon"],
    [PREPOSITION, "sous"],
    [PREPOSITION, "sous"],
    [PREPOSITION, "sur"],
    [PREPOSITION, "tour"],
    [PREPOSITION, "vers"],
    [PREPOSITION, "via"],
    // [PREPOSITION, "à cause de"],
    // [PREPOSITION, "à côté de"],
    // [PREPOSITION, "à côté de"],
    // [PREPOSITION, "à l'exception de"],
    // [PREPOSITION, "à l'extérieur"],
    // [PREPOSITION, "à l'intérieur"],
    // [PREPOSITION, "à partir de"],
    // [PREPOSITION, "à travers"],
    // [PREPOSITION, "au cours de"],
    // [PREPOSITION, "au lieu de"],
    // [PREPOSITION, "au nom de"],
    // [PREPOSITION, "au sommet de"],
    // [PREPOSITION, "aussi bien que"],
    // [PREPOSITION, "autant que"],
    // [PREPOSITION, "autour de"],
    // [PREPOSITION, "contrairement à"],
    // [PREPOSITION, "en dehors des"],
    // [PREPOSITION, "en dépit de"],
    // [PREPOSITION, "en dessous"],
    // [PREPOSITION, "en face de"],
    // [PREPOSITION, "en plus de"],
    // [PREPOSITION, "en raison de"],
    // [PREPOSITION, "jusqu'à ce que"],
    // [PREPOSITION, "l'intérieur de"],
    // [PREPOSITION, "loin d'être"],
    // [PREPOSITION, "près de"],
    // [PREPOSITION, "près de"],
    // [PREPOSITION, "près de"],
    // [PREPOSITION, "vers le bas"],

    [CONJUNCTION, "car"],
    [CONJUNCTION, "comme"],
    [CONJUNCTION, "donc"],
    [CONJUNCTION, "ensuite"],
    [CONJUNCTION, "et"],
    [CONJUNCTION, "lorsque"],
    [CONJUNCTION, "mais"],
    [CONJUNCTION, "ni"],
    [CONJUNCTION, "or"],
    [CONJUNCTION, "ou"],
    [CONJUNCTION, "plus"],
    [CONJUNCTION, "pourquoi"],
    [CONJUNCTION, "puis"],
    [CONJUNCTION, "puisque"],
    [CONJUNCTION, "quand"],
    [CONJUNCTION, "que"],
    [CONJUNCTION, "quoique"],
    [CONJUNCTION, "si"],
    // [CONJUNCTION, "à condition que"],
    // [CONJUNCTION, "à mesure que"],
    // [CONJUNCTION, "à moins que"],
    // [CONJUNCTION, "à supposer que"],
    // [CONJUNCTION, "afin que"],
    // [CONJUNCTION, "ainsi que"],
    // [CONJUNCTION, "alors que"],
    // [CONJUNCTION, "après que"],
    // [CONJUNCTION, "au cas où"],
    // [CONJUNCTION, "aussitôt que"],
    // [CONJUNCTION, "avant que"],
    // [CONJUNCTION, "bien que"],
    // [CONJUNCTION, "c’est-à-dire que"],
    // [CONJUNCTION, "comme si"],
    // [CONJUNCTION, "dans l’hypothèse où"],
    // [CONJUNCTION, "dans la mesure où"],
    // [CONJUNCTION, "de crainte que"],
    // [CONJUNCTION, "de façon que"],
    // [CONJUNCTION, "de manière que"],
    // [CONJUNCTION, "de même que"],
    // [CONJUNCTION, "de peur que"],
    // [CONJUNCTION, "de sorte que"],
    // [CONJUNCTION, "depuis que"],
    // [CONJUNCTION, "dès que"],
    // [CONJUNCTION, "en admettant que"],
    // [CONJUNCTION, "en attendant que"],
    // [CONJUNCTION, "encore que"],
    // [CONJUNCTION, "jusqu’à ce que"],
    // [CONJUNCTION, "même si"],
    // [CONJUNCTION, "ou bien"],
    // [CONJUNCTION, "parce que"],
    // [CONJUNCTION, "pendant que"],
    // [CONJUNCTION, "pour que"],
    // [CONJUNCTION, "pourvu que"],
    // [CONJUNCTION, "quand bien même"],
    // [CONJUNCTION, "quoi que"],
    // [CONJUNCTION, "sans que"],
    // [CONJUNCTION, "selon que"],
    // [CONJUNCTION, "sitôt que"],
    // [CONJUNCTION, "suivant que"],
    // [CONJUNCTION, "supposé que"],
    // [CONJUNCTION, "tandis que"],
    // [CONJUNCTION, "tant que"],
    // [CONJUNCTION, "une fois que"],
    // [CONJUNCTION, "vu que"],
    // [CONJUNCTION, "et ... et"],
    // [CONJUNCTION, "ni ... ni"],
    // [CONJUNCTION, "ou ... ou"],
    // [CONJUNCTION, "ou bien ... ou bien"],
    // [CONJUNCTION, "plus ... plus"],
    // [CONJUNCTION, "soit ... soit"],

    [PARTICLE, "ne"],
    [PARTICLE, "ô"],
]


const CHAR_CLASSES: [CharClass, string][] = [
    // acute accent
    [VOWEL, 'É'],
    [VOWEL, 'é'],
    // grave accent
    [VOWEL, 'À'],
    [VOWEL, 'È'],
    [VOWEL, 'Ù'],
    [VOWEL, 'à'],
    [VOWEL, 'è'],
    [VOWEL, 'ù'],
    // circumflex
    [VOWEL, 'Â'],
    [VOWEL, 'Ê'],
    [VOWEL, 'Î'],
    [VOWEL, 'Ô'],
    [VOWEL, 'Û'],
    [VOWEL, 'â'],
    [VOWEL, 'ê'],
    [VOWEL, 'î'],
    [VOWEL, 'ô'],
    [VOWEL, 'û'],
    // diaeresis
    [VOWEL, 'Ë'],
    [VOWEL, 'Ï'],
    [VOWEL, 'Ü'],
    [VOWEL, 'Ÿ'],
    [VOWEL, 'ë'],
    [VOWEL, 'ï'],
    [VOWEL, 'ü'],
    [VOWEL, 'ÿ'],
    // cedilla
    [CONSONANT, 'Ç'],
    [CONSONANT, 'ç'],
    // tilde
    [CONSONANT, 'Ñ'],
    [CONSONANT, 'ñ'],
    // ligatures
    [VOWEL, 'Æ'],
    [VOWEL, 'æ'],
    [VOWEL, 'Œ'],
    [VOWEL, 'œ'],
    [VOWEL, 'Ø'],
    [VOWEL, 'ø'],
]


const UTF_COMPOSE_MAP: [string, string][] = [
    // acute accent
    ["É", "É"],
    ["é", "é"],
    // grave accent
    ["À", "À"],
    ["È", "È"],
    ["Ù", "Ù"],
    ["à", "à"],
    ["è", "è"],
    ["ù", "ù"],
    // circumflex
    ["Â", "Â"],
    ["Ê", "Ê"],
    ["Î", "Î"],
    ["Ô", "Ô"],
    ["Û", "Û"],
    ["â", "â"],
    ["ê", "ê"],
    ["î", "î"],
    ["ô", "ô"],
    ["û", "û"],
    // diaeresis
    ["Ë", "Ë"],
    ["Ï", "Ï"],
    ["Ü", "Ü"],
    ["Ÿ", "Ÿ"],
    ["ë", "ë"],
    ["ï", "ï"],
    ["ü", "ü"],
    ["ÿ", "ÿ"],
    // cedilla
    ["Ç", "Ç"],
    ["ç", "ç"],
    // tilde
    ["Ñ", "Ñ"],
    ["ñ", "ñ"],
]


const UTF_REDUCE_MAP: [string, string][] = [
    // acute accent
    ["É", "E"],
    ["é", "e"],
    // grave accent
    ["À", "A"],
    ["È", "E"],
    ["Ù", "U"],
    ["à", "a"],
    ["è", "e"],
    ["ù", "u"],
    // circumflex
    ["Â", "A"],
    ["Ê", "E"],
    ["Î", "I"],
    ["Ô", "O"],
    ["Û", "U"],
    ["â", "a"],
    ["ê", "e"],
    ["î", "i"],
    ["ô", "o"],
    ["û", "u"],
    // diaeresis
    ["Ë", "E"],
    ["Ï", "I"],
    ["Ü", "U"],
    ["Ÿ", "Y"],
    ["ë", "e"],
    ["ï", "i"],
    ["ü", "u"],
    ["ÿ", "y"],
    // cedilla
    ["Ç", "C"],
    ["ç", "c"],
    // tilde
    ["Ñ", "N"],
    ["ñ", "n"],
    // ligatures
    ["Æ", "AE"],
    ["æ", "ae"],
    ["Œ", "OE"],
    ["œ", "oe"],
    ["Ø", "OE"],
    ["ø", "oe"],

]


export const LANG_FRENCH = new Lang(
    new FrenchStemmer(),
    [...CHAR_CLASSES_LATIN, ...CHAR_CLASSES],
    FUNCTION_WORDS,
    UTF_COMPOSE_MAP,
    UTF_REDUCE_MAP,
)
