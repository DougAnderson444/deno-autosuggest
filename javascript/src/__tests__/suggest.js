import {LucidSuggest, highlight} from "../index"
import {LANG_ENGLISH} from "../lang/lang-english"


describe("Suggest", () => {
    const records = [
        {id: 10, title: "Hello world!"},
        {id: 20, title: "Foo bar"},
        {id: 30, title: "-BAZZZ-"},
    ]

    const suggest = new LucidSuggest(LANG_ENGLISH)
    suggest.addRecords(records)

    test("Setup calls are sequential", async () => {
        function timeout(ms) {
            return new Promise((resolve) => {
                setTimeout(() => resolve(), ms)
            })
        }
        const steps = []
        suggest.setup(async () => {
            steps.push("start 1")
            await timeout(50)
            steps.push("end 1")
        })
        suggest.setup(async () => {
            steps.push("start 2")
            await timeout(50)
            steps.push("end 2")
        })
        await suggest.setupQueue()
        expect(steps).toEqual(["start 1", "end 1", "start 2", "end 2"])
    })

    test("Cannot search after destruction", async () => {
        const suggest = new LucidSuggest(LANG_ENGLISH)
        await suggest.addRecords(records)
        await suggest.search("foo")
        await suggest.destroy()
        await expect(suggest.search("foo")).rejects.toThrow("Suggest destroyed")
    })

    test("Empty input", async () => {
        const hits = await suggest.search("")
        expect(hits).toMatchSnapshot()
    })

    test("Equality", async () => {
        const hits = await suggest.search("foo bar")
        expect(hits).toMatchSnapshot()
    })

    test("Prefix", async () => {
        const hits = await suggest.search("ba")
        expect(hits).toMatchSnapshot()
    })

    test("Typos", async () => {
        const hits = await suggest.search("helo")
        expect(hits).toMatchSnapshot()
    })

    test("Typos", async () => {
        const hits = await suggest.search("helo")
        expect(hits).toMatchSnapshot()
    })

    test("Joined words", async () => {
        const hits = await suggest.search("foobar")
        expect(hits).toMatchSnapshot()
    })

    test("Limit", async () => {
        const hits1 = await suggest.search("b")
        expect(hits1).toHaveLength(2)

        await suggest.setLimit(1)
        const hits2 = await suggest.search("b")
        expect(hits2).toHaveLength(1)

        await suggest.setLimit(100)
        const hits3 = await suggest.search("b")
        expect(hits3).toHaveLength(2)
    })

    test("Rating", async () => {
        const suggest1 = new LucidSuggest(LANG_ENGLISH)
        suggest1.addRecords(records.map((r, i) => ({...r, rating: i})))
        const hits1 = await suggest1.search("")
        expect(hits1).toMatchSnapshot()

        const suggest2 = new LucidSuggest(LANG_ENGLISH)
        suggest2.addRecords(records.map((r, i) => ({...r, rating: records.length - i})))
        const hits2 = await suggest2.search("")
        expect(hits2).toMatchSnapshot()
    })

    describe("Render", () => {
        test("highlight helper", async () => {
            const suggest = new LucidSuggest(LANG_ENGLISH)
            suggest.addRecords(records)
            const hits = await suggest.search("hel")
            const rendered = hits.map(hit => highlight(hit, "((", "))"))
            expect(rendered).toMatchSnapshot()
        })

        test("DIY", async () => {
            const suggest = new LucidSuggest(LANG_ENGLISH)
            suggest.addRecords(records)
            const hits = await suggest.search("ba")
            const rendered = hits.map(hit => {
                return {
                    value: hit.record.id.toString(),
                    label: hit.chunks
                        .map(c => c.highlight ? "<<" + c.text + ">>" : c.text)
                        .join("")
                }
            })
            expect(rendered).toMatchSnapshot()
        })
    })
})