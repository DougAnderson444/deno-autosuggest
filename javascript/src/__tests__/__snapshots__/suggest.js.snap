// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggest Empty input 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "-BAZZZ-",
      },
    ],
    "record": Object {
      "id": 30,
      "title": "-BAZZZ-",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Foo bar",
      },
    ],
    "record": Object {
      "id": 20,
      "title": "Foo bar",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Hello world!",
      },
    ],
    "record": Object {
      "id": 10,
      "title": "Hello world!",
    },
  },
]
`;

exports[`Suggest Equality 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": true,
        "text": "Foo",
      },
      Object {
        "highlight": false,
        "text": " ",
      },
      Object {
        "highlight": true,
        "text": "bar",
      },
    ],
    "record": Object {
      "id": 20,
      "title": "Foo bar",
    },
  },
]
`;

exports[`Suggest Joined words 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": true,
        "text": "Foo",
      },
      Object {
        "highlight": false,
        "text": " ",
      },
      Object {
        "highlight": true,
        "text": "bar",
      },
    ],
    "record": Object {
      "id": 20,
      "title": "Foo bar",
    },
  },
]
`;

exports[`Suggest Prefix 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Foo ",
      },
      Object {
        "highlight": true,
        "text": "ba",
      },
      Object {
        "highlight": false,
        "text": "r",
      },
    ],
    "record": Object {
      "id": 20,
      "title": "Foo bar",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "-",
      },
      Object {
        "highlight": true,
        "text": "BA",
      },
      Object {
        "highlight": false,
        "text": "ZZZ-",
      },
    ],
    "record": Object {
      "id": 30,
      "title": "-BAZZZ-",
    },
  },
]
`;

exports[`Suggest Rating 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "-BAZZZ-",
      },
    ],
    "record": Object {
      "id": 30,
      "rating": 2,
      "title": "-BAZZZ-",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Foo bar",
      },
    ],
    "record": Object {
      "id": 20,
      "rating": 1,
      "title": "Foo bar",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Hello world!",
      },
    ],
    "record": Object {
      "id": 10,
      "rating": 0,
      "title": "Hello world!",
    },
  },
]
`;

exports[`Suggest Rating 2`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Hello world!",
      },
    ],
    "record": Object {
      "id": 10,
      "rating": 3,
      "title": "Hello world!",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "Foo bar",
      },
    ],
    "record": Object {
      "id": 20,
      "rating": 2,
      "title": "Foo bar",
    },
  },
  Hit {
    "chunks": Array [
      Object {
        "highlight": false,
        "text": "-BAZZZ-",
      },
    ],
    "record": Object {
      "id": 30,
      "rating": 1,
      "title": "-BAZZZ-",
    },
  },
]
`;

exports[`Suggest Render DIY 1`] = `
Array [
  Object {
    "label": "Foo <<ba>>r",
    "value": "20",
  },
  Object {
    "label": "-<<BA>>ZZZ-",
    "value": "30",
  },
]
`;

exports[`Suggest Render highlight helper 1`] = `
Array [
  "((Hel))lo world!",
]
`;

exports[`Suggest Typos 1`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": true,
        "text": "Hello",
      },
      Object {
        "highlight": false,
        "text": " world!",
      },
    ],
    "record": Object {
      "id": 10,
      "title": "Hello world!",
    },
  },
]
`;

exports[`Suggest Typos 2`] = `
Array [
  Hit {
    "chunks": Array [
      Object {
        "highlight": true,
        "text": "Hello",
      },
      Object {
        "highlight": false,
        "text": " world!",
      },
    ],
    "record": Object {
      "id": 10,
      "title": "Hello world!",
    },
  },
]
`;
